src/libraries/Microsoft.Extensions.Http/src/HttpClientFactoryOptions.cs

```diff
@@ -103,5 +109,58 @@ public TimeSpan HandlerLifetime
 
         internal bool SuppressDefaultLogging { get; set; }
         internal List<Action<HttpMessageHandlerBuilder>> LoggingBuilderActions { get; } = new List<Action<HttpMessageHandlerBuilder>>();
+
+        private List<Action<HttpMessageHandlerBuilder>>? _httpMessageHandlerBuilderActions;
+        private List<Action<IPrimaryHandlerBuilder>>? _primaryHandlerActions = new List<Action<IPrimaryHandlerBuilder>>();
+        private List<Action<IAdditionalHandlersBuilder>>? _additionalHandlersActions = new List<Action<IAdditionalHandlersBuilder>>();
+
+        internal IReadOnlyList<Action<IPrimaryHandlerBuilder>> PrimaryHandlerActions => _primaryHandlerActions!;
+        internal IReadOnlyList<Action<IAdditionalHandlersBuilder>> AdditionalHandlersActions => _additionalHandlersActions!;
+        internal bool MergedToHandlerBuilderActions => _primaryHandlerActions is null;
+
+        // fallback for backwards compatibility
+        private List<Action<HttpMessageHandlerBuilder>> MergeToHandlerBuilderActions()
+        {
+            Debug.Assert(_httpMessageHandlerBuilderActions is null);
+            Debug.Assert(_primaryHandlerActions is not null);
+            Debug.Assert(_additionalHandlersActions is not null);
+
+            List<Action<HttpMessageHandlerBuilder>> handlerBuilderActions = new List<Action<HttpMessageHandlerBuilder>>();
+            handlerBuilderActions.AddRange(_primaryHandlerActions);
+            handlerBuilderActions.AddRange(_additionalHandlersActions);
+
+            _primaryHandlerActions = null;
+            _additionalHandlersActions = null;
```

Comment:
Have these lists been handed out to anyone externally? If not, rather than allocating a new list and throwing away both existing ones, you could do:
```C#
List<Action<HttpMessageHandlerBuilder>> handlerBuilderActions = _primaryHandlerActions;
handlerBuilderActions .AddRange(_additionalHandlersActions);
_primaryHandlerActions = null;
_additionalHandlersActions = null;
return handlerBuilderActions;
```